### PR TITLE
Feature/swift rewrite

### PR DESCRIPTION
**Pull request by simonmitchell**
This is a fairly hefty PR, and may be best viewed in a git client. It's another step towards removing all Objective-C classes across our whole codebase!

### CODE REVIEW

**Comment by simonmitchell** [comment]
@MattCheetham @ryanbourneuk Would ideally like to get this merged before Matt leaves us 😢

File: ThunderBasics/SingleRequestLocationManager.swift [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
+        
+        guard let firstLocation = locations.first else { return }
+        
+        os_log("New location: %@", log: log, type: .debug, firstLocation)
+        callAllCompletionHandlersWith(location: firstLocation, error: nil)
+    }
+    
+    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
+        os_log("Location manager errored: %{public}@", log: log, type: .error, error.localizedDescription)
```
Comment (ryanbourneuk) [inline ThunderBasics/SingleRequestLocationManager.swift:160]: Where does `public` come from on this line? :)

File: ThunderBasics/ToastView.swift [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    private var gravity: UIGravityBehavior?
+    
+    private var completion: (() -> Void)?
+    
+    /// Shows the toast notification on the screen
+    ///
+    /// - Parameter completion: A completion block to be called when the toast notification has displayed and dismissed successfully
+    func show(completion: @escaping () -> Void) {
+        
+        frame = CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 44) // Add on 40 points so when it drops down you can't see the view behind it.
```
Comment (ryanbourneuk) [inline ThunderBasics/ToastView.swift:94]: Is the comment meant to say 4 instead of 40?

File: ThunderBasics/ToastView.swift [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        containerView.addSubview(self)
+        
+        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(handleTap(_:)))
+        coverWindow?.addGestureRecognizer(tapGesture)
+        
+        animator = UIDynamicAnimator(referenceView: containerView)
+        
+        gravity = UIGravityBehavior(items: [self])
+        gravity?.gravityDirection = CGVector(dx: 0.0, dy: 1.0)
+        animator?.addBehavior(gravity!)
```
Comment (ryanbourneuk) [inline ThunderBasics/ToastView.swift]: `gravity` is force-unwrapped here

File: ThunderBasics/ToastView.swift [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            guard let self = self else { return }
+            self.dismiss()
+        }
+    }
+    
+    @objc private func handleTap(_ sender: UITapGestureRecognizer) {
+        action?(self)
+    }
+    
+    private func layout() {
```
Comment (ryanbourneuk) [inline ThunderBasics/ToastView.swift:150]: A few hardcoded values here, could be good to define them as constants or add docs for them :)

File: ThunderBasics/UIColor+Helpers.swift [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+        var alpha: CGFloat = 0.0
+        
+        guard getRed(&red, green: &green, blue: &blue, alpha: &alpha) else {
+            return nil
+        }
+        
+        return (red, green, blue, alpha)
+    }
+    
+    /// Returns the RGBA components of the colour
```
Comment (ryanbourneuk) [inline ThunderBasics/UIColor+Helpers.swift]: Copy pasta here? :)

File: ThunderBasics/UIImage+ColorAnalysis.swift [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+//
+// Modified by David Schimon 1 / 2013
+// Modified by Simon Mitchell 02 / 10 / 2018
+
+import UIKit
+
+//MARK: - ImageColorAnalyzer -
+
+/// `ImageColorAnalyzer` Analyzes a `UIImage` and picks out its main colors
+public class ImageColorAnalyzer {
```
Comment (ryanbourneuk) [inline ThunderBasics/UIImage+ColorAnalysis.swift:21]: Would be quite good to get some performance tests around the functions here and see how long it'll take to execute with different images.

File: ThunderBasics/UIImage+Resize.swift [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            origin: CGPoint(
+                x: -(resizedImage.size.width - thumbnailSize)/2,
+                y: -(resizedImage.size.height - thumbnailSize)/2),
+            size: resizedImage.size
+        ))
+        
+        // Get the cropped image
+        let croppedImage = UIGraphicsGetImageFromCurrentImageContext()
+        
+        // Ket's gorget about that we were drawing
```
Comment (ryanbourneuk) [inline ThunderBasics/UIImage+Resize.swift]: `Let's forget` perhaps? :)

File: ThunderBasics/SingleRequestLocationManager.swift (hunk 1 above)
Comment (simonmitchell) [reply ThunderBasics/SingleRequestLocationManager.swift:160]: See the docs under "Privacy" :) https://developer.apple.com/documentation/os/logging

**Comment by simonmitchell** [comment]
@ryanbourneuk have made requested changes when you get a chance to give this another look-over! :D

**Review by ryanbourneuk — APPROVED** [review]
✅

**Pull request merged**